src/core/painting/qgsgeometrypaintdevice.cpp

```diff
@@ -0,0 +1,538 @@
+        break;
+      }
+
+      case QPainterPath::LineToElement:
+      {
+        double tx, ty;
+        matrix.map( e.x, e.y, &tx, &ty );
+        currentX << tx;
+        currentY << ty;
+        break;
+      }
+
+      case QPainterPath::CurveToElement:
+      {
+        Q_ASSERT( path.elementAt( i + 1 ).type == QPainterPath::CurveToDataElement );
+        Q_ASSERT( path.elementAt( i + 2 ).type == QPainterPath::CurveToDataElement );
+
+        const double x1 = path.elementAt( i - 1 ).x;
+        const double y1 = path.elementAt( i - 1 ).y;
+
+        double tx1, ty1;
+        matrix.map( x1, y1, &tx1, &ty1 );
+
+        double tx2, ty2;
+        matrix.map( e.x, e.y, &tx2, &ty2 );
+
+        const double x3 = path.elementAt( i + 1 ).x;
+        const double y3 = path.elementAt( i + 1 ).y;
+
+        double tx3, ty3;
+        matrix.map( x3, y3, &tx3, &ty3 );
+
+        const double x4 = path.elementAt( i + 2 ).x;
+        const double y4 = path.elementAt( i + 2 ).y;
+
+        double tx4, ty4;
+        matrix.map( x4, y4, &tx4, &ty4 );
+
+        // TODO -- we could likely reduce the number of segmented points here!
+        std::unique_ptr< QgsLineString> bezier( QgsLineString::fromBezierCurve(
+            QgsPoint( tx1, ty1 ),
+            QgsPoint( tx2, ty2 ),
+            QgsPoint( tx3, ty3 ),
+            QgsPoint( tx4, ty4 ) ) );
+
+        currentX << bezier->xVector();
+        currentY << bezier->yVector();
+
+        i += 2;
+        break;
+      }
+      case QPainterPath::CurveToDataElement:
+        Q_ASSERT( !"addSubpathGeometries(), bad element type" );
+        break;
+    }
+  }
+
+  if ( currentX.size() > 1 )
+  {
+    std::unique_ptr< QgsLineString > line = std::make_unique< QgsLineString >( currentX, currentY );
+    if ( line->isClosed() )
+    {
+      queuedPolygons.emplace_back( std::make_unique< QgsPolygon >( line.release() ) );
+    }
+    else
+    {
+      collection.addGeometry( line.release() );
+    }
+  }
+
+  if ( queuedPolygons.empty() )
+    return;
+
+  collection.reserve( collection.numGeometries() + queuedPolygons.size() );
+
+  QgsMultiPolygon tempMultiPolygon;
+  tempMultiPolygon.reserve( queuedPolygons.size() );
+  for ( auto &part : queuedPolygons )
+  {
+    tempMultiPolygon.addGeometry( part.release() );
+  }
+
+  // ensure holes are holes, not overlapping polygons
+  QgsGeos geosCollection( &tempMultiPolygon );
+  std::unique_ptr< QgsAbstractGeometry > g = geosCollection.makeValid( Qgis::MakeValidMethod::Linework );
+  if ( !g )
+    return;
+
+  for ( auto it = g->const_parts_begin(); it != g->const_parts_end(); ++it )
+  {
+    collection.addGeometry( ( *it )->clone() );
+  }
+}
+
+void QgsGeometryPaintEngine::drawPath( const QPainterPath &path )
+{
+  QPainterPath realPath = path;
+  if ( mUsePathStroker )
+  {
+    QPen pen = painter()->pen();
+    QPainterPathStroker stroker( pen );
+    QPainterPath strokedPath = stroker.createStroke( path );
+    realPath = strokedPath;
+  }
```

Comment:
This is meant to not return and continue onward with the addSubpathGeometries() functio right?

Reply:
Correct -- we stroke the path then fallback to the same logic for rendering the polygon path

Reply:
@nyalldawson , approved.